mondey_backend/src/mondey_backend/routers/statistics.py

```diff
@@ -0,0 +1,399 @@
+        max_age_months = 72
+        count = np.zeros(max_age_months + 1, dtype=np.int32)
+        avg = np.zeros(max_age_months + 1, dtype=np.float64)
+        stddev = np.zeros(max_age_months + 1, dtype=np.float64)
+
+    if child_ages == {}:
+        return count, avg, stddev
+
+    # online algorithm computes variance, compute m2 from stddev
+    # we can ignore count-1 <= 0 because stddev is zero in this case
+    m2 = stddev**2 * (count - 1)
+
+    for answer in answers:
+        age = child_ages[answer.answer_session_id]  # type: ignore
+        new_count, new_avg, new_m2 = _add_sample(
+            count[age], avg[age], m2[age], answer.answer + 1
+        )
+        count[age] = new_count
+        avg[age] = new_avg
+        m2[age] = new_m2
+
+    count, avg, stddev = _finalize_statistics(count, avg, m2)  # type: ignore
+
+    return count, avg, stddev
+
+
+def calculate_milestone_statistics_by_age(
+    session: SessionDep, milestone_id: int, session_expired_days: int = 7
+) -> MilestoneAgeScoreCollection | None:
+    """
+    Calculate the mean, variance of a milestone per age in months.
+    Takes into account only answers from expired sessions. If no statistics exist yet, all answers from expired sessions are considered, else only the ones newer than the last statistics are considered.
+
+    Parameters
+    ----------
+    session : SessionDep
+        database session
+    milestone_id : int
+        id of the milestone to calculate the statistics for
+    session_expired_days : int, optional
+        expiration age of an answersession, by default 7
+
+    Returns
+    -------
+    MilestoneAgeScoreCollection | None
+        MilestoneAgeScoreCollection object which contains a list of MilestoneAgeScore objects,
+    one for each month, or None if there are no answers for the milestoneg and no previous statistics.
+    """
+    # get the newest statistics for the milestone
+    last_statistics = session.exec(
+        select(MilestoneAgeScoreCollection)
+        .where(col(MilestoneAgeScoreCollection.milestone_id) == milestone_id)
+        .order_by(col(MilestoneAgeScoreCollection.created_at).desc())
+    ).first()
+
+    # initialize avg and stddev scores with the last known statistics or to None if no statistics are available
+    child_ages = _get_answer_session_child_ages_in_months(session)
+    expiration_date = datetime.datetime.now() - datetime.timedelta(
+        days=session_expired_days
+    )
+
+    count = None
+    avg_scores = None
+    stddev_scores = None
+
+    if last_statistics is None:
+        # no statistics exists yet -> all answers from expired sessions are relevant
+
+        # README: type: ignore here because mypy flags the second argument, but the documentation of sqlquery has
+        # boolean conditions: https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.join
+        # the code-docstring of this function has it as example as well: selectable.py
```

Comment:
hm, ok. it's a little weird because if I follow the documentation of the `join` function it looks like this: 
```python 
r"""Create a SQL JOIN against this :class:`_expression.Select`
        object's criterion
        and apply generatively, returning the newly resulting
        :class:`_expression.Select`.

        E.g.::

            stmt = select(user_table).join(address_table, user_table.c.id == address_table.c.user_id)

        The above statement generates SQL similar to::ini120458
        

            SELECT user.id, user.name FROM user JOIN address ON user.id = address.user_id
"""
``` 
which doesn't have a `col`, nor parentheses like the official documentation does.  anyway, it works with `col`...

Reply:
right - that's (I assume?) the sqlalchemy docs - and we're using SQLModel which is sqlalchemy+pydantic, and here we want just the sqlalchemy column from the SQLModel object attribute, hence the `col()`

Reply:
ok... this distinction confuses me all the time...